Simple GATKConf (NO MERGE)

magicDGS opened the pull request:
Very simple implementation of #2297 using a custom `GATKConf` class to allow both promatically (`GATKConfBuilder`) and by commons-configuration API (constructor). Only includes:

* Packages/Classes to include in the CLP on startup.
* Packages to look for codecs on startup.

magicDGS commented:
could some of you have a look to this, @droazen and/or @lbergelson? We started the discussion in #2156, so maybe this implementation is a good starting point.

droazen commented on src/main/java/org/broadinstitute/hellbender/Main.java:
```diff
@@ -100,7 +92,9 @@ public Object instanceMain(final String[] args, final List<String> packageList,
     public Object instanceMain(final String[] args) {
-        return instanceMain(args, getPackageList(), getClassList(), getCommandLineName());
+        final GATKConf configuration = getConfiguration();
+        FeatureManager.setConfiguration(configuration);
```
Rather than pass the full configuration object to engine classes like `FeatureManager`, I'd prefer to pass just the `List<String>` of packages to use for codec discovery.

magicDGS replied:
Done.

magicDGS replied:
I also move this part of the code to the other `instanceMain()` method, passing just the configuration object generated here instead of the list to set the `CommandLineProgram` configuration properly.

droazen commented on src/main/java/org/broadinstitute/hellbender/Main.java:
```diff
@@ -57,20 +61,8 @@
-    protected List<Class<? extends CommandLineProgram>> getClassList() {
-        return Collections.emptyList();
+    protected GATKConf getConfiguration() {
+        return GATKConfBuilder.getDefaultConfiguration();
```
I think it would be better if the master configuration object lived as a field in `CommandLineProgram` accessible by subclasses via an inherited method. This would simplify passing property-based information around the engine in the future.

magicDGS replied:
I agree partially on this. It is better to have `GATKConf` in `CommandLineProgram`, but if this method remains, to allow extending `Main` classes to set their own configuration on startup. This is important for one reason: if an API client wants its own configuration and it only lives in `CommandLineProgram` accessible by subclasses, it requires that every time a tool is implemented this method should be overridden. Thus, it makes more dirty to implement walker types which only requires an apply, for example.

magicDGS replied:
I added a settable field in `CommandLineProgram` in a similar way as in `FeatureManager`.

droazen commented on src/main/java/org/broadinstitute/hellbender/conf/GATKConf.java:
```diff
@@ -0,0 +1,58 @@
+     * The single classes we wish to include in our command line.
+     */
+    @SuppressWarnings("unchecked")
+    public List<Class<? extends CommandLineProgram>> getToolClasses() {
```
I think configuration properties should always be typed as Strings. Otherwise we have to worry about typing issues / casting, etc. Even in this simple branch, there are already inconsistencies in typing (eg., depending on whether you call `setClpClasses()` or `addClpClasses()`, the clpClasses might be Strings or Classes

magicDGS replied:
The inconsistency is my fault, because I haven't paid too much attention in the builder (I'm sorry, I detected other inconsistencies).

I think that typing as String is a good idea, even more if the `GATKConf` use constants to access the properties.

magicDGS replied:
As before, I removed this method.

droazen commented on src/main/java/org/broadinstitute/hellbender/conf/GATKConf.java:
```diff
@@ -0,0 +1,58 @@
+     * The packages we wish to include in our command line.
+     */
+    @SuppressWarnings("unchecked")
+    public List<String> getToolPackages() {
```
I'm not sure we want to have to define a separate method for each property we add. Perhaps we could just have a bunch of constants with the property names, as you do above, and have the client just pass in one of those constants when accessing properties.

magicDGS replied:
I did it to cast the list properly in the case of the tool classes, for instance. It is true that as the required configuration increases, this class will grow a lot.

magicDGS replied:
I removed these methods and create `GATKConf.getGATKProperty(String key)` instead.

droazen commented on src/main/java/org/broadinstitute/hellbender/conf/GATKConfBuilder.java:
```diff
@@ -0,0 +1,92 @@
+    public GATKConfBuilder setDefaults() {
+        return setToolsPackages(new ArrayList<>(DEFAULT_TOOLS_PACKAGES))
+                .setClpClasses(new ArrayList<>())
+                .setCodecPackages(new ArrayList<>(DEFAULT_CODEC_PACKAGES));
```
Can we check in a properties file with the default settings that gets loaded on startup, instead of hardcoding them here?

magicDGS replied:
A properties file will be great, but I decided to hardcode these properties because how is it going to be handled this file in project dependencies? I'm not sure how to do this in a proper way, because I'm not coming from the IT background...I may need some help on this.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureManager.java:
```diff
@@ -42,34 +45,33 @@
+     * @throws IllegalStateException if codecs were alredy initialized.
      */
-    static {
+    public synchronized static void setConfiguration(final GATKConf configuration) {
```
Replace with `setCodecPackages()`, called from `CommandLineProgram` using the information from the master configuration object.

magicDGS replied:
Done.

magicDGS replied:
But I call it in `instanceMain(final String[] args, final GATKConf configuration, final String commandLineName)`.

magicDGS commented:
I addressed some of your comments, @droazen. If you would like to have a properties file for the configuration, I will need some help on setting it up (although I will try by my own too).

Although I still set up some of the environment in `Main`, now the `CommandLineProgram` class have the same instance passed by `Main`.

Looking forward for your comments on the updates.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/conf/GATKConf.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * We will search these packages at startup to look for FeatureCodecs.
+     */
+    public final static String CODEC_PACKAGE_LIST_KEY = "codecPackages";
```
Another package list required is the one to look for read filters in `GATKReadFilterPluginDescriptor`.

cmnbroad commented:
One proposal for moving forward would be to have a default properties file with a known name/location that is included in the gatk jar (say, "gatk.default.properties"), which is always loaded and populates the initial configuration, and then use the classloader getResources method to also load all resources with some other known name (say, "gatk.properties"). That way any properties files on the classpath with the known name would be automatically discovered and loaded. The apache commons API allows looks like it has good support for handling this using a [composite](http://commons.apache.org/proper/commons-configuration/userguide/howto_compositeconfiguration.html#Composite_Configuration_Details) configuration. We would have to define some rules around override semantics, but it looks like the api provides a lot of control over that as well.

magicDGS commented:
Where should be sited "gatk.default.properties"? I'm still investigating a bit on how to do this in the proper way...

droazen commented:
Closing -- this is supplanted by the work being done in https://github.com/broadinstitute/gatk/issues/2368. @jonn-smith You should continue to use this closed PR as a reference for understanding @magicDGS 's use case.
